### ISSUE TITLE
convnet performance

### ISSUE THREAD

**Issue by alucantonio** [gorgonia/gorgonia#378]
I am considering using Gorgonia for a research project, as an alternative to Tensorflow, since I have other code already written in Go. I tried to run the `convnet` example included in Gorgonia: it takes almost 25 s per epoch on my i7-7th gen machine running Arch Linux, while a similar example in Tensorflow (https://www.tensorflow.org/tutorials/keras/classification?hl=en) takes about 3 s per epoch.  If I am correct, the two examples are comparable. I am not using the GPU because in Gorgonia I am experiencing issue #376. Do you have any idea about the cause of Gorgonia's poor performance?

**Comment by chewxy** [comment]
Hah. This is not supposed to be the case - the `convnet` example was built for benchmarking against other systems. This means that some changes made so far have degraded the performance in general. Please hold while I investigate.

**Comment by owulveryck** [comment]
Which type do you use? Float32 of float64?

**Comment by alucantonio** [comment]
@owulveryck I tried both Float32 and Float64 and there does not seem to be any substantial difference in the computation time per epoch.

**Comment by chewxy** [comment]
Working on several CUDA related issues today, will have a quick look on this to figure out where it's gone wrong.

**Comment by alucantonio** [comment]
Ok, let me know if I can help.

**Comment by chewxy** [comment]
I can confirm that currently Gorgonia is slower than Keras. I also found a few bugs in the `losses`, so fixing those first. On CUDA, the speeds are equivalent.

Will keep you updated

**Comment by chewxy** [comment]
Hi, a bunch of fixes just went in.. will be benchmarking this in a bit

**Comment by chewxy** [comment]
v0.9.10 if anyone is following along

**Comment by alucantonio** [comment]
Great, thanks! I am downloading the new version right now.

**Comment by alucantonio** [comment]
On my machine the `convnet` example still takes about 2.5 minutes per epoch, as before the update to  v0.9.10. It is still pretty slow: can anyone confirm this?

By the way, I think that on line 219 in `main.go` I believe that
`cost := gorgonia.Must(gorgonia.Neg(cost))`
should be replaced by
`cost := gorgonia.Must(gorgonia.Neg(losses))`

**Comment by chewxy** [comment]
Hmm, ok looking into it

**Comment by chewxy** [comment]
More vectorized instructions were added to `vecf32` and `vecf64`. Will have a few more updates to those packages

**Comment by alucantonio** [comment]
Has anyone tried comparing any benchmark with TensorFlow or other Python libraries? I would like to understand if it is an issue with my platform or it is really Gorgonia's current performance.

(Maybe there is another place where we can have this discussion)

**Comment by blackrez** [comment]
@alucantonio 
Can you profile your golang code with [pprof](https://golang.org/pkg/runtime/pprof/) ?
Also what is your platform ?

There is also a [Slack channel](https://gophers.slack.com/messages/gorgonia/).
Thanks.

**Comment by alucantonio** [comment]
@blackrez Thanks for the answer. Here is the output of the `top` command in pprof:
`File: main`
`Type: cpu`
`Time: May 6, 2020 at 1:58pm (CEST)`
`Duration: 2.48mins, Total samples = 2.49mins (100.37%)`
`Entering interactive mode (type "help" for commands, "o" for options)`
`(pprof) top`
`Showing nodes accounting for 122.80s, 82.34% of 149.13s total`
`Dropped 228 nodes (cum <= 0.75s)`
`Showing top 10 nodes out of 78`
`      flat  flat%   sum%        cum   cum%`
`    31.99s 21.45% 21.45%     31.99s 21.45%  gonum.org/v1/gonum/internal/asm/f64.DotUnitary`
`    20.87s 13.99% 35.45%     20.88s 14.00%  gorgonia.org/gorgonia.im2colOp.f64s`
`    12.68s  8.50% 43.95%     12.68s  8.50%  gorgonia.org/tensor.(*FlatIterator).ndNext`
`    12.43s  8.34% 52.28%     12.78s  8.57%  gorgonia.org/gorgonia.(*maxPoolOp).f64s`
`    11.26s  7.55% 59.83%     11.26s  7.55%  runtime.memclrNoHeapPointers`
`     8.68s  5.82% 65.65%     40.67s 27.27%  gonum.org/v1/gonum/blas/gonum.dgemmSerialNotTrans`
`     7.16s  4.80% 70.46%     29.17s 19.56%  gorgonia.org/gorgonia.StandardEngine.Transpose`
`     6.30s  4.22% 74.68%     18.98s 12.73%  gorgonia.org/tensor.(*FlatIterator).Next`
`     5.73s  3.84% 78.52%      5.73s  3.84%  runtime.memmove`
`     5.70s  3.82% 82.34%      5.70s  3.82%  gorgonia.org/tensor/internal/execution.GteSameVSF64`

The profiling refers to **one epoch of training of the convnet example**. As you can see, it takes 2.48 minutes.

Here is the info on my machine:
CPU: Intel i7-7700HQ @ 2.80GHz
RAM: 16 GB
OS: Arch Linux - 4.19.119 kernel
I have the latest versions (last commits on master) of gonum, gorgonia and the other dependencies.

To join the Slack channel I think I need an invitation: how can I have it?

**Comment by blackrez** [comment]
Yes, you can have one here : https://invite.slack.golangbridge.org/.

**Comment by owulveryck** [comment]
@alucantonio your pprof shows that the most consuming task is the actual matrix multiplication involved in the Convolution operator.
This is weird because the implementation heavily relies on the gonum implementation of the basic operations sgemm. There is no reason why this operation would be significantly faster on another library (maybe slightly faster, but not ten times)
